From 66d72f79f559374209ad2bd1783c1066eab91161 Mon Sep 17 00:00:00 2001
From: Matt Madison <matt@madison.systems>
Date: Thu, 15 Jan 2026 00:40:20 -0800
Subject: [PATCH 2/2] feat: capsule updates with single-boot configuration

Let the EFI boot manager find ESP partitions, even for a single-boot
setup, so that capsule updates can be applied. Have the PlatformBootOrder
library prepend the single-boot application to the boot order, so the ESP
does not need to contain a boot launcher.

Signed-off-by: Matt Madison <matt@madison.systems>
---
 .../PlatformBootManagerLib/PlatformBm.c       |  4 ++
 .../PlatformBootOrderLib.c                    | 50 +++++++++++++++----
 .../PlatformBootOrderLib.inf                  |  1 +
 .../PlatformBootOrderLibNoIpmi.inf            |  3 ++
 4 files changed, 48 insertions(+), 10 deletions(-)

diff --git a/Silicon/NVIDIA/Library/PlatformBootManagerLib/PlatformBm.c b/Silicon/NVIDIA/Library/PlatformBootManagerLib/PlatformBm.c
index 91f8089e..620f0df5 100644
--- a/Silicon/NVIDIA/Library/PlatformBootManagerLib/PlatformBm.c
+++ b/Silicon/NVIDIA/Library/PlatformBootManagerLib/PlatformBm.c
@@ -2145,6 +2145,10 @@ PlatformBootManagerBeforeConsole (
     if (!EFI_ERROR (Status)) {
       mRscHandler->Register (SingleBootStatusCodeCallback, TPL_CALLBACK);
     }
+
+    EfiBootManagerRefreshAllBootOption ();
+    SetBootOrder ();
+
   }
 
   //
diff --git a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.c b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.c
index 5c94b837..7235d9c3 100644
--- a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.c
+++ b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.c
@@ -13,6 +13,7 @@
 #include <Library/MemoryAllocationLib.h>
 #include <Library/HobLib.h>
 #include <Library/PlatformResourceLib.h>
+#include <Library/PcdLib.h>
 #include <Library/PrintLib.h>
 #include <Library/UefiBootManagerLib.h>
 #include <Library/UefiLib.h>
@@ -47,6 +48,9 @@ NVIDIA_BOOT_ORDER_PRIORITY  mBootPriorityTemplate[BOOT_ORDER_TEMPLATE_CLASS_COUN
 STATIC  NVIDIA_BOOT_ORDER_PRIORITY  *mBootPriorityTable = NULL;
 STATIC  UINTN                       mBootPriorityCount  = 0;
 
+STATIC NVIDIA_BOOT_ORDER_PRIORITY mBootAppTemplate =
+{ "bootapp",    0, MAX_UINT8,             MAX_UINT8,             MAX_UINT8,                MAX_UINTN, MAX_UINTN, MAX_UINTN, MAX_UINTN };
+
 STATIC
 EFI_PCI_IO_PROTOCOL *
 GetBootOptPciIoProtocol (
@@ -334,6 +338,16 @@ GetBootClassOfOption (
   Bus      = MAX_UINTN;
   Device   = MAX_UINTN;
   Function = MAX_UINTN;
+
+  if (FeaturePcdGet (PcdSingleBootSupport) && StrCmp (Option->Description, L"Boot Application") == 0) {
+    for (BootPriorityIndex = 0; BootPriorityIndex < Count; BootPriorityIndex++) {
+      if (AsciiStrCmp (Table[BootPriorityIndex].OrderName, "bootapp") == 0) {
+        Result = &Table[BootPriorityIndex];
+        goto ReturnResult;
+      }
+    }
+  }
+
   if (StrCmp (Option->Description, L"UEFI Shell") == 0) {
     for (BootPriorityIndex = 0; BootPriorityIndex < Count; BootPriorityIndex++) {
       if (AsciiStrCmp (Table[BootPriorityIndex].OrderName, "shell") == 0) {
@@ -637,17 +651,11 @@ STATIC
 INTN
 EFIAPI
 BootOrderSortCompare (
-  IN CONST VOID  *Buffer1,
-  IN CONST VOID  *Buffer2
+  IN CONST EFI_BOOT_MANAGER_LOAD_OPTION  *Left,
+  IN CONST EFI_BOOT_MANAGER_LOAD_OPTION  *Right
   )
 {
-  INT32  Priority1;
-  INT32  Priority2;
-
-  Priority1 = GetDevicePriority (*(UINT16 *)Buffer1);
-  Priority2 = GetDevicePriority (*(UINT16 *)Buffer2);
-
-  return Priority1 - Priority2;
+  return GetDevicePriority (Left->OptionNumber) - GetDevicePriority (Right->OptionNumber);
 }
 
 STATIC
@@ -673,7 +681,29 @@ ParseDefaultBootPriority (
   UINTN                       BootPrioritySbdfLen;
   UINTN                       BootClassIndex;
 
+  // For single-boot setup, make sure the single-boot app is first
   Priority = 0;
+  if (FeaturePcdGet (PcdSingleBootSupport)) {
+    NVIDIA_BOOT_ORDER_PRIORITY  *NewBuffer;
+    NewBuffer = (NVIDIA_BOOT_ORDER_PRIORITY *)ReallocatePool (
+                                                mBootPriorityCount * sizeof (NVIDIA_BOOT_ORDER_PRIORITY),
+                                                (mBootPriorityCount + 1) * sizeof (NVIDIA_BOOT_ORDER_PRIORITY),
+                                                (VOID *)mBootPriorityTable
+                                              );
+    if (NewBuffer == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a: Failed to allocate boot table buffer\r\n", __FUNCTION__));
+    } else {
+      mBootPriorityTable = NewBuffer;
+      CopyMem (
+        &mBootPriorityTable[mBootPriorityCount],
+        &mBootAppTemplate,
+        sizeof (mBootAppTemplate)
+      );
+      ClassBootPriority = &mBootPriorityTable[mBootPriorityCount++];
+      ClassBootPriority->PriorityOrder = Priority++;
+    }
+  }
+
   // Process the priority order
   Status = GetVariable2 (
              L"DefaultBootPriority",
@@ -808,7 +838,7 @@ SetBootOrder (
 
   ParseDefaultBootPriority ();
 
-  EfiBootManagerSortLoadOptionVariable (LoadOptionTypeBoot, BootOrderSortCompare);
+  EfiBootManagerSortLoadOptionVariable (LoadOptionTypeBoot, (SORT_COMPARE)BootOrderSortCompare);
 
   VariableData = TRUE;
   gRT->SetVariable (
diff --git a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.inf b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.inf
index cdcc828d..960bf07c 100644
--- a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.inf
+++ b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLib.inf
@@ -43,6 +43,7 @@
 
 [Pcd]
   gNVIDIATokenSpaceGuid.PcdIpmiNetworkBootMode
+  gNVIDIATokenSpaceGuid.PcdSingleBootSupport
 
 [Guids]
   gEfiGlobalVariableGuid
diff --git a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLibNoIpmi.inf b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLibNoIpmi.inf
index e2ea7d68..d3fd0405 100644
--- a/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLibNoIpmi.inf
+++ b/Silicon/NVIDIA/Library/PlatformBootOrderLib/PlatformBootOrderLibNoIpmi.inf
@@ -39,6 +39,9 @@
 [Protocols]
   gEfiPciIoProtocolGuid          ## CONSUMES
 
+[Pcd]
+  gNVIDIATokenSpaceGuid.PcdSingleBootSupport
+
 [Guids]
   gEfiGlobalVariableGuid
   gNVIDIATokenSpaceGuid
-- 
2.43.0

